Nez.Portable/UI/Stage.cs

```diff
@@ -168,105 +169,154 @@ public void update()
 			if( _isGamepadFocusEnabled )
 				updateGamepadState();
 			updateKeyboardState();
+			updateInputMouse();
 
+			#if !FNA
+			if( Input.touch.isConnected && Input.touch.currentTouches.Count > 0 )
+			{
+				updateInputTouch();
+			}
+			#endif
+		}
+
+
+		/// <summary>
+		/// Handle mouse input events.
+		/// </summary>
+		void updateInputMouse()
+		{
 			// consolidate input checks so that we can add touch input easily later
-			var leftMouseButtonDown = Input.leftMouseButtonDown;
-			var leftMouseButtonPressed = Input.leftMouseButtonPressed;
-			var leftMouseButtonReleased = Input.leftMouseButtonReleased;
 			var currentMousePosition = getMousePosition();
 
-			var didMouseMove = false;
+			var mouseMoved = false;
 			if( _lastMousePosition != currentMousePosition )
 			{
-				didMouseMove = true;
+				mouseMoved = true;
 				_lastMousePosition = currentMousePosition;
 			}
 
+			var inputPos = screenToStageCoordinates( currentMousePosition );
+
+			updateInputPoint( inputPos, Input.leftMouseButtonPressed, Input.leftMouseButtonReleased,
+				mouseMoved, ref _mouseOverElement );
+		}
+
+
+		/// <summary>
+		/// Handle all the touch input events.
+		/// </summary>
+		void updateInputTouch()
+		{
 			#if !FNA
-			// convert touch to mouse
-			if( Input.touch.isConnected && Input.touch.currentTouches.Count > 0 )
+			foreach( var touch in Input.touch.currentTouches )
 			{
-				var touch = Input.touch.currentTouches[0];
-				leftMouseButtonDown = touch.State == Microsoft.Xna.Framework.Input.Touch.TouchLocationState.Pressed || touch.State == Microsoft.Xna.Framework.Input.Touch.TouchLocationState.Moved;
-				leftMouseButtonPressed = touch.State == Microsoft.Xna.Framework.Input.Touch.TouchLocationState.Pressed;
-				leftMouseButtonReleased = touch.State == Microsoft.Xna.Framework.Input.Touch.TouchLocationState.Released || touch.State == Microsoft.Xna.Framework.Input.Touch.TouchLocationState.Invalid;
-				currentMousePosition = touch.Position;
-
-				didMouseMove = touch.State == Microsoft.Xna.Framework.Input.Touch.TouchLocationState.Moved;
-				if( didMouseMove )
-					_lastMousePosition = currentMousePosition;
+				var inputPos = screenToStageCoordinates( touch.Position );
+				var inputPressed = touch.State == Microsoft.Xna.Framework.Input.Touch.TouchLocationState.Pressed;
+				var inputReleased = touch.State == Microsoft.Xna.Framework.Input.Touch.TouchLocationState.Released || touch.State == Microsoft.Xna.Framework.Input.Touch.TouchLocationState.Invalid;
+				var inputMoved = false;
+				Microsoft.Xna.Framework.Input.Touch.TouchLocation prevTouch;
+				if( touch.TryGetPreviousLocation( out prevTouch ) )
+				{
+					if( Vector2.Distance( touch.Position, prevTouch.Position ) >= float.Epsilon )
+						inputMoved = true;
+				}
+				Element lastOver;
+				_touchOverElement.TryGetValue( touch.Id, out lastOver );
+
+				updateInputPoint( inputPos, inputPressed, inputReleased, inputMoved, ref lastOver );
+
+				_touchOverElement[touch.Id] = lastOver;
```

Comment:
If I am following the code correctly, _touchOverElement is never nulled out on mouse/touch up. In most cases that would be fine but for a dynamic UI something like this could occur:

- UI changes adding 5 buttons on screen
- user touches each button with each finger, so we now have the _touchOverElement dict with 5 Elements in it
- user releases all touches. The _touchOverElement dict is still holding strong references to the previously touched Elements.
- dynamic UI changes and the buttons are removed but they wont be cleaned up by the GC because of the _touchOverElement references.